Composer/packages/client/src/pages/language-generation/code-editor.tsx

```diff
@@ -2,45 +2,108 @@
 // Licensed under the MIT License.
 
 /* eslint-disable react/display-name */
-import React, { useState, useEffect } from 'react';
-import { LgEditor } from '@bfc/code-editor';
+import React, { useState, useEffect, useMemo, useContext } from 'react';
+import { LgEditor, LGOption } from '@bfc/code-editor';
 import get from 'lodash/get';
 import debounce from 'lodash/debounce';
 import isEmpty from 'lodash/isEmpty';
-import { CodeRange, LgFile } from '@bfc/shared';
+import { Diagnostic } from 'botbuilder-lg';
+import { LgFile } from '@bfc/shared';
 
+import { StoreContext } from '../../store';
 import * as lgUtil from '../../utils/lgUtil';
 
 interface CodeEditorProps {
   file: LgFile;
-  onChange: (value: string) => void;
-  codeRange?: Partial<CodeRange> | null;
+  template: lgUtil.Template | null;
 }
 
+// lsp server port should be same with composer/server port.
+const lspServerPort = process.env.NODE_ENV === 'production' ? process.env.PORT || 3000 : 5000;
```

Comment:
If this code belongs on client-side (browser), it shouldn't be able to access `NODE_ENV`.

Reply:
William we use a webpack plugin to serialize process.env into the client. No sensitive information is expected.

Reply:
William we use a webpack plugin to serialize process.env into the client. No sensitive information is expected.